[Dart 3] Records (#4766)
dart-lang/site-www: src/language/records.md

---
title: Records
description: Summary of the record data structure in Dart.
toc: false
---

Records are an anonymous, immutable, aggregate type. Like other [collection types][], 
they let you bundle multiple objects into a single object. Unlike other collection 
types, records are fixed-sized, heterogeneous, and typed.

Records are real values; you can store them in variables, pass them to and from

> functions, and store them in lists.

Reviewer: This is a great intro.

Author: I'm pretty sure you wrote that sentence 😅 definitely a great intro!

## Record syntax

Records are comma-delimited lists of named or positional fields,
enclosed in parenthesis:

```dart
var record = ('first', a: 2, b: true, 'last');    // Record expression
```

Named fields inside record type annotations go inside a brace-delimited section
of type and name pairs:

```dart

> (Sring, {int a, bool b}, String?) record;    // Record type annotation

Author: Nice, didn't know that!

> ```

Reviewer: The comments in the code samples help, but I worry this is still too subtle of a jump from a record *expression* to a record *type annotation*. Because it jumps from positional fields to named fields, it's easy to overlook that it's also moving from an expression to a type annotation too.

I'd probably insert an example of a record type annotation with only positional fields first and explain that it's a type annotation and where you might use it.

Reviewer: For what it's worth, I always use "type annotation" to mean the syntax for how to write down a type. They aren't optional "annotations" in the Dart 1 sense, but as far as I know, "type annotation" is pretty standard across languages as a way to refer to type terms.

Author: > I'd probably insert an example of a record type annotation with only positional fields first and explain that it's a type annotation _and where you might use it._

@munificent Where _would_ you use one? 
Do they only come up for destructuring? 

I'm drawing a blank [from the spec](https://github.com/dart-lang/language/blob/master/accepted/future-releases/records/records-feature-specification.md#record-type-annotations). I was going to say, "to declare a record", but that doesn't sound right, since they're immutable it's not like you can declare a record then populate it later (can you?)

[Even for typing](https://github.com/dart-lang/language/blob/master/accepted/future-releases/records/records-feature-specification.md#record-type-annotations:~:text=(num%2C%20num)%20r%20%3D%20(3%2C%203.5)..%241.isEven%3B), it seems like the types are going to be inferred regardless of what you write in a type annotation, so that doesn't seem like a use case, just an example for the sake of the spec.

Reviewer: Yes, you rarely need them for variable declarations because they can be inferred. But they come up at API boundaries: Parameter types and return types. For example, here's a function that takes a record and swaps its fields:

```dart
(int, int) swap((int, int) record) {
  var (a, b) = record;
  return (b, a);
}
```

The two `(int, int)` here are record type annotations.

> Naming fields in record type annotations is optional for documentation purposes.

Reviewer: The names of positional fields are optional and only for documentation, but the names of named fields (I admit that's confusing) are not optional. So these two record types are equivalent as far as the type checker is concerned:

```dart
(int x, int y, int z) point = (1, 2, 3);
(int r, int g, int b) color = (1, 2, 3);

// OK:
point = color;
```

But these are not:

```dart
({int x, int y, int z}) point = (x: 1, y: 2, z: 3);
({int r, int g, int b}) color = (r: 1, g: 2, b: 3);

// Error, different types:
point = color;
```

(This is the same for function parameters, for what it's worth.)

Author: Oh ok, thanks for explaining. I might just use this example to illustrate that point.

Author: Just realized, the exact opposite of what I wrote in the next line 😅 good thing you caught that!

Author: @munificent Actually, I'm still not sure I understand. In the first example, if I were to say `print(point.x)` or `print(color.b)`, it would work and print `1` and `3`, right?

And what if you were to write:
```
(int x, int y, int z) point = (a: 1, b: 2, c: 3);
```
That wouldn't work, right?

Author: Nevermind, you can ignore all of this. I was somehow reading "the names _of named_ fields" as just "the names of fields" (which was very confusing!), but now I get it.

The names don't need to match any record expression corresponding to the type.

## Record fields

Record fields are accessible through built-in getters. Records are immutable,
so fields do not have setters. 

Named fields expose getters of the same name. Positional fields expose getters
of the name `$<position>`, skipping named fields:

```dart

> print(record.$1);    // Prints '123'

Reviewer: Might be worth repeating the declaration of `record` here in case it's not clear that it's coming from the previous example.

print(record.a);     // Prints 2
print(record.b);     // Prints true
print(record.$2);    // Prints 'last'
```

To streamline record field access even more, see the page on [Patterns][].

{% comment %}
    TODO: link to patterns page, specifically records destructuring section.
{% endcomment %}

## Record types

> There is no explicit record type; records are

Reviewer: It's probably more accurate to say that there are a plethora of record types, one for each unique set of field names, positions, and field types. So `(int, int)` is a record type, `({x: int, y: int})` is another record type, etc.

Also, all of those record types inherit from a built in type named `Record`. So there sort of is "an" explicit record type, though `Record` isn't useful for much. (It's directly analogous to `Function` and function types.)

I would maybe word this like:

> Records are structurally typed: the set of the fields, their types, and their names, uniquely determines the type of a record. Two records with the same set of field names and types have the same type.

Author: Is it accurate to say a records shape is "the the set of its fields, the fields' types, and their names (if any)"? I use "shape" later on in the page so it seems like this would be a good place to define it first

structurally typed based on the types of their fields.

Each field in a record has its own type, which can be different from other
fields' types in the same record. Records are transparent to the type system.
The type system remembers and applies each fields' type wherever it is
accessed from the record:

```dart
var oneField = record.$1;       // oneField is a String
var anotherField = record.a;    // anotherField is an int 
```

If two unrelated libraries create records with the same set of fields,
the type system understands that those records are the same type even though the

> libraries are not coupled to each other.

Reviewer: 👍

Reviewer 2: And *that* is where structural types differ from nominative types. :+1: indeed!

## Record equality

Two records are equal if they have the same _shape_ (set of fields),
and their corresponding fields have the same values.
Named field order does not affect a record's shape:

```dart
var a = (x: 1, 2);
var b = (2, x: 1);
assert(a == b);        // True
```

Records automatically define `hashCode` and `==` methods structurally based on
their fields.

{% comment %}
    Example of record as a composite hash key in map? Supposedly a benefit of
    record value semantice.
{% endcomment %}

## Multiple returns

Records allow functions to return multiple values bundled together:

```dart
(String, int) userInfo(Map<String, dynamic> json) {
  return (json['name'] as String, json['age'] as int);
}

> main() {

Reviewer 2: I'd add `void ` in front of `main`. Users get yelled at by lints if they forget the return type of top-level functions, so we should be setting a good example (when it doesn't cost more than that).

If anything, I think readers will visually pattern-match `void main() {` more easily than just `main() {`, because they are more used to the former.

Author: Very good to know!

  var info = userInfo(json);
  var name = info.$1;
  var age  = info.$2;
}

> ```

Reviewer 2: I'd risk showing an example of pattern matching here, as an appetite wetter for linking to patterns.
After all, we intend pattern matching to be primary way to destructure a record, so I think we should mention it in this document too. Maybe even prominently. Record destructuring using patterns *is* a record feature (too).

Maybe:
> or, using the recommended pattern destructuring approach:
> ```dart
> void main() {
>   var (name, age) = userInfo(json);
> }
> ```

> The only ways to accomplish multiple returns otherwise are

Reviewer 2: I'd drop "only" here. Maybe "existing". And I'd drop "otherwise" too, it doesn't read well where it's placed, and the sentence works without it if using "existing".

Author: I think "existing" is dating the documentation too much; it'll sound weird once records have also existed for a while. But I get what you're saying, I think "other" will suffice.

either to create a class (which is verbose), or use

> another collection type like List or Map (which lose type safety).

Reviewer 2: I'd use code quotes on the source names: "like \`List\` or \`Map\`"

> another collection type like List or Map (which lose type safety).

Reviewer 2: lose -> loses  

Maybe?
(I read "which" as referring to the singular act of "using something". Consistent with how "which is verbose" uses the singular "is".)

{{site.alert.note}}
  Records' multiple-return and heterogenous-type characteristics enable

>   parallelization of Futures of different types, which you can read about in the

Author: I was wondering about this! I saw a lot of non-code-quoted upper-cased "Future"s in the existing section of the library tour, and then some non-code-quoted _lower_-cases, and code-quoted `Future`s in other places. So, any rhyme or reason is much appreciated 😅

---
title: Records
description: Summary of the record data structure in Dart.
toc: false
---

Records are an anonymous, immutable, aggregate type. Like other [collection types][], 
they let you bundle multiple objects into a single object. Unlike other collection 
types, records are fixed-sized, heterogeneous, and typed.

Records are real values; you can store them in variables, pass them to and from
functions, and store them in lists.

## Record syntax

_Records expressions_ are comma-delimited lists of named or positional fields,
enclosed in parenthesis:

```dart
var record = ('first', a: 2, b: true, 'last');    // Record expression
```

Named fields inside _record type annotations_ go inside a brace-delimited section
of type and name pairs, after all positional fields:

```dart
(String, String?, {int a, bool b}) record;    // Record type annotation
```
Naming fields in record type annotations is optional for documentation purposes,

> but the names of named fields are not optional. For more information,

Author: I think it means, once given a name in annotation, the name can be omitted but it can't be replaced. Changed this to

```
Naming fields in record type annotations is optional for documentation purposes.
The names of those named fields are not optional, though,  
once you create a record with named fields. The field names become part of the
[record's type definition](#record-types). They can be omitted, but not replaced.

For more information, see the [Record equality](#record-equality) section.
```

fields' types in the same record. The type system is aware of each
field's type wherever it is accessed from the record:

```dart
(num, Object) pair = (42, "a");

var first = pair.$1;     // static type `num`, runtime type `int`
var second = pair.$2;    // static type `Object`, runtime type `String`
```

If two unrelated libraries create records with the same set of fields,
the type system understands that those records are the same type even though the
libraries are not coupled to each other.

## Record equality

Two records are equal if they have the same _shape_ (set of fields),
and their corresponding fields have the same values.
Since named field order is not part of a record's shape, the order of named
fields does not affect equality:

```dart
(int x, int y, int z) point = (1, 2, 3);
(int r, int g, int b) color = (1, 2, 3);

// OK:
point = color;
```

```dart
({int x, int y, int z}) point = (x: 1, y: 2, z: 3);
({int r, int g, int b}) color = (r: 1, g: 2, b: 3);

// Error, different types:
point = color;
```

Records automatically define `hashCode` and `==` methods structurally based on
their fields.

## Multiple returns

> Records allow functions to return multiple values bundled together. Using [pattern matching][], you can retrieve record values from a return by destructuring them directly into local variables:

Reviewer 3: Split into multiple lines?

{% comment %}
    TODO: link to patterns page, specifically records destructuring section.
{% endcomment %}

```dart
// Returns multiple values in a record:
(String, int) userInfo(Map<String, dynamic> json) {
  return (json['name'] as String, json['age'] as int);
}

void main() {
  // Destructures using a record pattern:
  var (name, age) = userInfo(json);

  /* Equivalent to:
  var info = userInfo(json);
  var name = info.$1;
  var age  = info.$2;
  */
}
```

Other ways to accomplish multiple returns are

> either to create a class (which is verbose), or use

Reviewer 3: ```suggestion
either to create a class (which is more verbose), or use
```

toc: false
---

Records are an anonymous, immutable, aggregate type. Like other [collection types][], 
they let you bundle multiple objects into a single object. Unlike other collection 
types, records are fixed-sized, heterogeneous, and typed.

Records are real values; you can store them in variables, pass them to and from
functions, and store them in lists.

## Record syntax

_Records expressions_ are comma-delimited lists of named or positional fields,
enclosed in parenthesis:

```dart
var record = ('first', a: 2, b: true, 'last');    // Record expression
```

_Record type annotations_ are comma-delimited lists of types enclosed in parenthesis.
The `(int, int)` statements here are record type annotations used as a return
type and as parameter types:

```dart
(int, int) swap((int, int) record) {
  var (a, b) = record;
  return (b, a);
}
```

Named fields inside _record type annotations_ go inside a brace-delimited section
of type and name pairs, after all positional fields:

```dart
// Record type annotation in a variable declaration
(String, String?, {int a, bool b}) record;
```

Naming fields in record type annotations is optional for documentation purposes.

> The names of those named fields are not optional, though,  
> once you create a record with named fields. The field names become part of the
> [record's type definition](#record-types). They can be omitted, but not replaced.

Reviewer 4: **issue:** Sentence structure per [GSG](https://developers.google.com/style/sentence-structure)
```suggestion
Once you create a record with named fields,
those field names become part of the [record's type definition](#record-types).
They can be omitted, but not replaced.
```

**question:**  If they're not optional, how can you omit the field names?

Author: It's confusing. How I understand it, which I haven't found a good way to put into words yet, is: once you name fields and actually create a record using those names (so, not just using a record type annotation to define its shape, but using a record expression to populate the fields), the names become part of the record's definition, i.e. it's shape. 

You can create a record from an annotation with named fields _without_ using the names, because they're implicit(maybe? not super sure about this part), but you _can't_ create a record from an annotation with named fields using different names in the expression...

```dart 
   ({int x, int y}) someRecord = (x: 1, y: 2);
// ^^^^^^^^^^^^^^^              ^^^^^^^^^^^^
// type annotation              expression

   ({int x, int y}) someRecord;
   someRecord = (1, 2);
// fine, field values fit the types in the annotation, and don't clash with the names
   print(someRecord.x);    // prints 1, i think?
   print(someRecord.y);    // prints 2, i think?

   ({int x, int y}) someRecord;
   someRecord = (a: 1, b: 2);
// doesn't work, can't give the fields different names because someRecord's shape is defined as named int x and named int y
```

So, something like that... but I'm not 100% sure that's accurate

Reviewer: I'll see if I can help:

You can think of every record as having two separate sets of fields:

* A set of positional fields.
* A set of named fields.

These two sets are totally independent of each other: a named field isn't a positional field and a positional field isn't a named field. Either set may be empty, or they can both be. You can have records with just positional fields, just named, both, or no fields at all.

### Record expressions

When you write a record expression to create a record object, the positional fields are written as simple expressions:

```dart
var record = (1, 2, 3);
//          ^  ^  ^ Positional fields.
```

The named fields have a name and `:` before the field value:

```dart
var record = (x: 1, y: 2, z: 3);
//          ^^^^  ^^^^  ^^^^ Named fields.
```

You can have both in a single record expression and they can be arbitrarily
interleaved:

```dart
var record = (x: 1, 2, y: 3, 4);
//                ^        ^  Positional fields.
//          ^^^^     ^^^^  Named fields.
```

### Record types

Every record object has a static type, which is how the type checker "sees" the record. Users write down these types using "type annotations" (which are not related to metadata annotations, the `@foo` syntax).

In a record type annotation, the positional fields are written as simple type annotations for the field types:

```dart
  (int, String) record = (1, 'str');
// ^^^  ^^^^^^ Positional fields.
```

Named fields are enclosed in curly braces and must have a name after the type:

```dart
  ({int a, String b}) record = (a: 1, b: 'str');
//  ^^^^^  ^^^^^^^^ Named fields.
```

If a record type annotation has both positional fields and named ones, the positional ones must come first:

```dart
  (double, bool, {int a, String b}) record = (1.2, true, a: 1, b: 'str');
// ^^^^^^^ ^^^^ Positional fields.
//                ^^^^^  ^^^^^^^^ Named fields.
```

### Positional field names

The above is, I hope, fairly clear and straightforward. Where it gets a little confusing is that we do allow names on positional fields *in record type annotations*:

```dart
  (int x, String y) record = (1, 'str');
// ^^^^^  ^^^^^^^^ Positional fields.
```

Note that the fields aren't enclosed in curly braces. These are still positional fields and are not named fields. Note how in the record expression in the initializer, there are no names. It's just that the fields in the type annotation have, uh, "names".

But these "names" are purely for documentation and have no meaning to the language or the type checker. It's essentially the same as doing:

```dart
  (int /* x */, String /* y */) record = (1, 'str');
// ^^^          ^^^^^^ Positional fields.
```

And, in particular, these names are *not* part of the record's actual type:

```dart
(int x, String y) record1 = (1, 'str');
(int a, String b) record2 = record1; // OK!
```

These two variables, `record1` and `record2` have the exact same type even though the positional field "names" are different, which is why the assignment is valid.

This is different from actual named fields where the name *is* part of the type:

```dart
({int x, String y}) record1 = (x: 1, y: 'str');
({int a, String b}) record2 = record1; // Error!
```

This assignment is invalid because `record1` and `record2` have different types. Note also that the record expression in the initializer now says the field names.

The proposal sometimes talks about the "name" of a positional field, but really we should have a different term for it because it's not the same as an actual named field's name.

Sorry this is confusing. Much of this comes from deliberately designing records to mirror function parameter lists, so we inherit some weirdness from that. But I think the overall correspondence is useful (especially if we ever add the ability to spread a record into an argument list).

You can see the exact same confusion around positional parameter names in function typedefs:

```dart
typedef Positional1 = Function(int x, String y);
typedef Positional2 = Function(int a, String b);
```

These two typedefs define equivalent types even though the positional parameters have different names. But these two are different types from each other:

```dart
typedef Named1 = Function({int x, String y});
typedef Named2 = Function({int a, String b});
```

> [record's type definition](#record-types). They can be omitted, but not replaced.

Reviewer 5: Wording seems confusing, maybe more clearly differentiate what is the difference between a "field with a name" and "named field"? Maybe mention that the first one concerned "positional fields"? Bob's analogy to function parameters seems good.

Author: Changed 43-45 to:

> Once you create a record using a record expression, though,
> any field names become part of the [record's type definition](#record-types)
> and are immutable. They can be omitted, but not replaced.

And then the link to the record equality section shows the example:

> ```dart
> // point =/= color
> ({int x, int y, int z}) point = (x: 1, y: 2, z: 3);
> ({int r, int g, int b}) color = (r: 1, g: 2, b: 3);
> ```

 is that better? Admittedly, I've been having trouble capturing this point, so any feedback helps

Reviewer 5: I'll mark this as resolved since you adopted munificent's suggestion

> For more information, see the [Record equality](#record-equality) section.

Reviewer 4: **nit:** Prefer different construction because there is no call to action for the reader.
```suggestion
To learn more, check out the [Record equality](#record-equality) section.
```

## Record fields

Record fields are accessible through built-in getters. Records are immutable,
so fields do not have setters. 

Named fields expose getters of the same name. Positional fields expose getters
of the name `$<position>`, skipping named fields:

```dart
var record = ('first', a: 2, b: true, 'last');

> print(record.$1);    // Prints '123'

Reviewer 5: Shouldnt this print `'first'`?

Reviewer 2: Yep, looks like a typo. There is no "123" to print.

print(record.a);     // Prints 2
print(record.b);     // Prints true
print(record.$2);    // Prints 'last'
```

To streamline record field access even more, see the page on [Patterns][].

{% comment %}
    TODO: link to patterns page, specifically records destructuring section.
{% endcomment %}

## Record types

> There is no type declaration for individual record types. Records are structurally
> typed based on the types of their fields. A record's _shape_ (the set of its fields,
> the fields' types, and their names, if any) uniquely determines the type of a record. 

Reviewer 4: **suggestion:** Use active voice per [GSG](https://developers.google.com/style/voice). Though the object doesn't need to own the action, adding agency helps here. 

```suggestion
Records don't declare the types for individual record types.
The record's _shape_ determines the specific type of a record. 
A shape includes the set of its fields, the fields' types, and any field names.
```

> Each field in a record has its own type, which can be different from other
> fields' types in the same record. The type system is aware of each
> field's type wherever it is accessed from the record:

Reviewer 4: **suggestion:** Active voice per [GSG](https://developers.google.com/style/voice). 
```suggestion
Each field in a record has its own type. Fields can differ from other fields' types in the same record.
The type system is aware of each field's type wherever the record accesses the type.
```

Author: Skipping second sentence rewrite, the record doesn't access itself, it could be any construct doing the accessing hence generic "wherever"

```dart
(num, Object) pair = (42, "a");

var first = pair.$1;     // static type `num`, runtime type `int`
var second = pair.$2;    // static type `Object`, runtime type `String`
```

> If two unrelated libraries create records with the same set of fields,
> the type system understands that those records are the same type even though the
> libraries are not coupled to each other.

Reviewer 4: **issue:** Write as shorter sentences per [GSG](https://developers.google.com/style/accessibility#ease-of-reading).
```suggestion
You can create records with the same field set in unrelated libraries.
Despite no link between the libraries, the type system infers that those records are the same type.
```

## Record equality

Two records are equal if they have the same _shape_ (set of fields),
and their corresponding fields have the same values.
Since named field _order_ is not part of a record's shape, the order of named

> fields does not affect equality:

Reviewer 6: this sentence ends with a colon but the example doesn't demonstrate the sentence's claim

```dart
(int x, int y, int z) point = (1, 2, 3);
(int r, int g, int b) color = (1, 2, 3);

// OK:
point = color;
```

```dart
({int x, int y, int z}) point = (x: 1, y: 2, z: 3);
({int r, int g, int b}) color = (r: 1, g: 2, b: 3);

// Error, different types:
point = color;
```

> Records automatically define `hashCode` and `==` methods structurally based on
> their fields.

Reviewer 4: **nits:** "Automatically" is redundant. The user can't manually define the hashCode, correct? "Structurally" reads as a hedge. 
```suggestion
Records define their `hashCode` and `==` methods based on the structure of their fields.
```

Author: Users can manually define hasCode and ==, so "automatically" actually is the selling point here. 

"based on the structure of their fields" sounds good

## Multiple returns

Records allow functions to return multiple values bundled together.

> Using [pattern matching][], you can retrieve record values from a return by
> destructuring them directly into local variables:

Reviewer 4: **issue:** Condition, then action per [GSG](https://developers.google.com/style/sentence-structure).
```suggestion
To retrieve record values from a return,
destructure the values into local variables using [pattern matching][].
```

{% comment %}
    TODO: link to patterns page, specifically records destructuring section.
{% endcomment %}

```dart
// Returns multiple values in a record:
(String, int) userInfo(Map<String, dynamic> json) {
  return (json['name'] as String, json['age'] as int);
}

void main() {
  // Destructures using a record pattern:
  var (name, age) = userInfo(json);

  /* Equivalent to:
  var info = userInfo(json);
  var name = info.$1;
  var age  = info.$2;
  */
}
```

> Other ways to accomplish multiple returns are
> either to create a class (which is more verbose), or use
> another collection type like `List` or `Map` (which loses type safety).

Reviewer 4: **issue:** Avoid parentheses per [GSG](https://developers.google.com/style/parentheses). Try to ease reading per [GSG](https://developers.google.com/style/accessibility#ease-of-reading).
```suggestion
You can accomplish multiple returns in other ways that have some downsides.
* Create a class. This is more verbose.
* Use another collection type like `List` or `Map`. This loses type safety.
```

Author: I don't want to use the bullet list because I feel like it draws too much attention to something that's just kind of a throwaway comment and not about records. Maybe see if the edit in the new commit is reasonable?

```dart
(int, int) swap((int, int) record) {
  var (a, b) = record;
  return (b, a);
}
```

Named fields inside _record type annotations_ mirror how [named parameters and arguments][]
work in functions. They go inside a curly brace-delimited
section of type-and-name pairs, after all positional fields:

```dart
// Record type annotation in a variable declaration:
(String, String?, {int a, bool b}) record;
```

Naming fields in record type annotations is optional for documentation purposes.
Once you create a record from a type annotation with named fields using a record
expression, the names become part of the [record's type definition](#record-types),
or its _shape_:

```dart 
  ({int x, int y}) someRecord = (x: 1, y: 2);
// ^^^^^^^^^^^^^^^              ^^^^^^^^^^^^
// type annotation              expression
```

They can be ommitted from the creation expression, but not replaced
with other names.

```dart
// Type annotation:
({int x, int y}) someRecord;
// Record expression creating record from type annotation, omitting names:
someRecord = (1, 2);
// Prints 1, 2:
print(someRecord.x, someRecord.y);
```

For more information and examples, check out [Record types](#record-types) and

> [Record equality](#record-equality).

Author: @munificent this was the only significant change from the last commit (lines 34-67), could you review it one more time? There were a couple comments on the confusing wording so I tried to elaborate

Reviewer: It's getting there, but still not quite accurate. How about:

---

Fields in record expressions and type annotations mirror how [parameters and arguments][] work in functions. Positional fields go directly inside the parentheses:

```dart
// Record type annotation in a variable declaration:
(String, int) record;

// Initialize it with a record expression:
record = ('A string', 123);
```

In a record type annotation, named fields go inside a curly brace-delimited
section of type-and-name pairs, after all positional fields. In a record
expression, the names go before each field value with a colon after:

```dart
// Record type annotation in a variable declaration:
({int a, bool b}) record;

// Initialize it with a record expression:
record = (a: 123, b: true);
```

The names of named fields in a record type are part of thee [record's type
definition](#record-types), or its _shape_. Two records with named fields with
different names have different types:

```dart
({int a, int b}) recordAB = (a: 1, b: 2);
({int x, int y}) recordXY = (x: 3, y: 4);

recordAB = recordXY; // Compile error! These records don't have the same type.
```

In a record type annotation, you can also name the *positional* fields, but
these names are purely for documentation and don't affect the record's type:

```dart
(int a, int b) recordAB = (1, 2);
(int x, int y) recordXY = (3, 4);

recordAB = recordXY; // OK.
```

This is similar to how positional parameters in a function declaration or function typedef can have names but those names don't affect the signature of the function.

For more information and examples, check out [Record types](#record-types) and
[Record equality](#record-equality).

Author: Thanks, I'll open another PR for this!

## Record syntax

_Records expressions_ are comma-delimited lists of named or positional fields,
enclosed in parentheses:

```dart
// Record expression:
var record = ('first', a: 2, b: true, 'last');
```

_Record type annotations_ are comma-delimited lists of types enclosed in parentheses.

> The `(int, int)` statements in the following example are record type annotations
> used as a return type and as parameter types:

Reviewer 4: **suggestion:** Rephrase and split to ease readability.
```suggestion
The following example includes the `(int, int)` statements
as record type annotations.
The statements work as both a return type and as parameter types:
```

}
```

Named fields inside _record type annotations_ mirror how [named parameters and arguments][]
work in functions. They go inside a curly brace-delimited
section of type-and-name pairs, after all positional fields:

```dart
// Record type annotation in a variable declaration:
(String, String?, {int a, bool b}) record;
```

> Naming fields in record type annotations is optional for documentation purposes.

Reviewer 4: **nit:** Write in active voice per [GSG](https://developers.google.com/style/voice).
```suggestion
Record type annotations don't require naming fields for documentation purposes.
```

Once you create a record from a type annotation with named fields using a record

> expression, the names become part of the [record's type definition](#record-types),
> or its _shape_:

Reviewer 4: **nit:** The comma isn't needed.
```suggestion
expression, the names become part of the [record's type definition](#record-types) or its _shape_:
```
